core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java

```diff
@@ -0,0 +1,118 @@
+import java.util.Map;
+
+public final class InnerHitsBuilder implements Writeable<InnerHitsBuilder>, ToXContent {
+
+    public final static InnerHitsBuilder PROTO = new InnerHitsBuilder(Collections.emptyMap());
+
+    private final Map<String, InnerHitBuilder> innerHitsBuilders;
+
+    public InnerHitsBuilder() {
+        this.innerHitsBuilders = new HashMap<>();
+    }
+
+    public InnerHitsBuilder(Map<String, InnerHitBuilder> innerHitsBuilders) {
+        this.innerHitsBuilders = innerHitsBuilders;
+    }
+
+    public InnerHitsBuilder addInnerHit(String name, InnerHitBuilder builder) {
+        this.innerHitsBuilders.put(name, builder.setName(name));
+        return this;
+    }
+
+    public Map<String, InnerHitBuilder> getInnerHitsBuilders() {
+        return innerHitsBuilders;
+    }
+
+    @Override
+    public InnerHitsBuilder readFrom(StreamInput in) throws IOException {
+        int size = in.readVInt();
+        Map<String, InnerHitBuilder> innerHitsBuilders = new HashMap<>();
+        for (int i = 0; i < size; i++) {
+            innerHitsBuilders.put(in.readString(), new InnerHitBuilder(in));
+        }
+        return new InnerHitsBuilder(innerHitsBuilders);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        for (Map.Entry<String, InnerHitBuilder> entry : innerHitsBuilders.entrySet()) {
+            builder.field(entry.getKey(), entry.getValue(), params);
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(innerHitsBuilders.size());
+        for (Map.Entry<String, InnerHitBuilder> entry : innerHitsBuilders.entrySet()) {
+            out.writeString(entry.getKey());
+            entry.getValue().writeTo(out);
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        InnerHitsBuilder that = (InnerHitsBuilder) o;
+        return innerHitsBuilders.equals(that.innerHitsBuilders);
+
+    }
+
+    @Override
+    public int hashCode() {
+        return innerHitsBuilders.hashCode();
+    }
+
+    public static InnerHitsBuilder fromXContent(XContentParser parser, QueryParseContext context) throws IOException {
+        Map<String, InnerHitBuilder> innerHitBuilders = new HashMap<>();
+        for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+            switch (token) {
+                case START_OBJECT:
+                    String name = parser.currentName();
+                    InnerHitBuilder innerHitBuilder = InnerHitBuilder.fromXContent(parser, context);
+                    innerHitBuilder.setName(name);
+                    innerHitBuilders.put(name, innerHitBuilder);
+                    break;
+            }
```

Comment:
Can we throw ParseExceptions on unexpected tokens here, e.g. when we encounter arrays or find anything that is not a START_OBJECT after a field name?